include/hlslib/xilinx/Simulation.h

```diff
@@ -62,13 +68,39 @@ class _Dataflow {
  public:
   inline static _Dataflow& Get() { // Singleton pattern
     static _Dataflow df;
-    return df; 
+    return df;
   }
 
   template <class Ret, typename... Args>
   void AddFunction(Ret (*func)(Args...), non_deducible_t<Args>... args) {
-    threads_.emplace_back(func, passed_by(std::forward<Args>(args),
-                                          std::is_reference<Args>{})...);
+    if (subflow_active_) {
+      subflows_.back().AddFunction(func, args...);
+    } else {
+      threads_.emplace_back(func, passed_by(std::forward<Args>(args),
+                                            std::is_reference<Args>{})...);
+    }
+  }
+
+  template <class Ret, typename... Args>
+  void AddSubflow(Ret (*func)(Args...), non_deducible_t<Args>... args) {
+    if (subflow_active_) {
+      subflows_.back().AddSubflow(func, args...);
+    } else {
+      subflows_.emplace();
+      subflow_active_ = true;
+      threads_.emplace_back(func, passed_by(std::forward<Args>(args),
+                                            std::is_reference<Args>{})...);
+      while (subflow_active_) continue;
```

Comment:
Spinning on an atomic variable is expensive. I think we should use a condition variable for this

Reply:
I can see how that could become a problem with deep hierarchies of subflows.
However, I have never used condition_variables and after trying for about 30 minutes and getting nothing but deadlocks I have given up (too much other stuff to do sadly).
If you have experience with them, could you try and implement it?

Reply:
Nevermind, I took some time and made it work.
I checked and it now indeed does not use CPU while waiting.

Reply:
Thank you! I can take it from here :-)

Sorry, I didn't realize that you were not familiar with condition variables. Maybe the time you invested will pay off in the future next time you need them!

Reply:
Always happy to learn something new! Deadlocks are just terribly frustrating...